lib/mobile_app_backend/pub_sub.ex

```diff
@@ -0,0 +1,120 @@
+  into the format expected by subscribers.
+  """
+  @type registry_value :: {Store.fetch_keys(), function()}
+
+  @doc """
+  Group registered pids by the data they are subscribed to.
+  """
+  @spec group_pids_by_target_data([{pid(), registry_value()}]) :: %{registry_value() => [pid()]}
+  def group_pids_by_target_data(registry_entries) do
+    Enum.group_by(
+      registry_entries,
+      fn {_, {fetch_keys, format_fn}} -> {fetch_keys, format_fn} end,
+      fn {pid, _} -> pid end
+    )
+  end
+
+  @doc """
+  Broadcast the latest data to subscriber pids if the data has changed since they last received it.
+  """
+  @spec broadcast_latest_data(any(), atom(), registry_value(), [pid()], atom()) :: any()
+  def broadcast_latest_data(
+        latest_data,
+        broadcast_message_name,
+        registry_value,
+        pids,
+        last_dispatched_table_name
+      ) do
+    last_dispatched_entry = :ets.lookup(last_dispatched_table_name, registry_value)
+
+    if !already_broadcast(last_dispatched_entry, latest_data) do
+      broadcast(
+        pids,
+        latest_data,
+        broadcast_message_name,
+        registry_value,
+        last_dispatched_table_name
+      )
+    end
+  end
+
+  defp already_broadcast([], _latest_data) do
+    # Nothing has been broadcast yet
+    false
+  end
+
+  defp already_broadcast([{_registry_key, old_data}], latest_data) do
+    old_data == latest_data
+  end
+
+  defp broadcast(
+         pids,
+         data,
+         broadcast_message_name,
+         {fetch_keys, _format_fn} = registry_value,
+         last_dispatched_table_name
+       ) do
+    Logger.info("#{__MODULE__} broadcasting to pids len=#{length(pids)}")
+
+    {time_micros, _result} =
+      :timer.tc(__MODULE__, :broadcast_to_pids, [
+        pids,
+        data,
+        broadcast_message_name
+      ])
+
+    Logger.info(
+      "#{__MODULE__} broadcast_to_pids fetch_keys=#{inspect(fetch_keys)} duration=#{time_micros / 1000}"
+    )
+
+    :ets.insert(last_dispatched_table_name, {registry_value, data})
+  end
+
+  def broadcast_to_pids(pids, data, broadcast_message_name) do
+    Enum.each(
+      pids,
+      &send(
+        &1,
+        {broadcast_message_name, data}
+      )
+    )
+  end
+
+  defmacro __using__(opts) do
+    quote location: :keep, bind_quoted: [opts: opts] do
+      use GenServer
+
+      broadcast_interval_ms = Keyword.fetch!(opts, :broadcast_interval_ms)
+
+      # Any time there is a reset_event, broadcast so that subscribers are immediately
+      # notified of the changes. This way, when the stream first starts,
+      # consumers don't have to wait `:broadcast_interval_ms` to receive their first message.
+      @impl true
+      def handle_info(:reset_event, state) do
+        send(self(), :broadcast)
+        {:noreply, state, :hibernate}
+      end
+
+      def handle_info(:timed_broadcast, state) do
+        send(self(), :broadcast)
+        interval = unquote(broadcast_interval_ms)
```

Comment:
I'm surprised that this `unquote` works, since `broadcast_interval_ms` is not a variable within the scope of `__using__/1` - I'd expect a bare `broadcast_interval_ms` to work here, although storing `broadcast_interval_ms` as a module attribute `@broadcast_interval_ms` might be closer to what we'd write by hand.

Reply:
I tried removing unquote but it doesn't work. Isn't `broadcast_interval_ms` defined as a variable in the scope of `__using__/1`/ [here](https://github.com/mbta/mobile_app_backend/pull/244/files#diff-f72fb54bb8f4d6600ef1e7dbdce4949fd1f712ad51295b7fe6097b2a25979198R97)?

I like the approach of passing it as a param so that it helps to direct a reader to see that the interval-based broadcasting logic is done by the macro, rather than having to otherwise search for usage of `@broadcast_interval_ms`

Reply:
Ah thanks for the explanation, I misunderstood your initial suggestion. Will update!